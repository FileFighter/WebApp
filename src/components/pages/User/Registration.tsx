import React, {ReactElement, useEffect, useState} from "react";
import {Alert, Container, Row} from "react-bootstrap";
import {notMinStrLength} from "../../../background/methods/checkInput";
import fileFighter from "../../../assets/images/logos/adventurer-run.gif";
import {registerNewUser} from "../../../background/api/registration";
import {getWindowSize, getWindowSize_Interface} from "../../../background/methods/windowSize";
import {getStyleValue} from "../../../background/methods/style";
import {DEFAULT_ALERT_DURATION, MIN_PASSWORD_LENGTH} from "../../../background/constants";
import UserInformationInput, {UserInformationInputInterface} from "./UserInformationInput";

export default function Registration(): ReactElement {
    const [alertMessage, setAlertMessage] = useState<string>("Error 404: No Message found.");
    const [alertVariant, setAlertColor] = useState<"primary" | "secondary" | "success" | "danger" | "warning" | "info" | "light" | "dark">("success");
    const [alertVisibility, setAlertVisibility] = useState<boolean>(false);

    const registrationContainer = document.getElementById("registrationContainer")
    const logoSubmit = document.getElementById("logoSubmit")

    useEffect(() => {
        function repositionSubmitLogo() {
            const logo = document.getElementById("logoSubmit")
            if (logo) {
                const container: HTMLElement | null = document.getElementById("registrationContainer");
                const leftContainerOffset: number = container?.getBoundingClientRect().left ?? 0;

                let containerPadding: string | number | null = getStyleValue(container, "padding-left");
                const pxPosition = containerPadding.indexOf("px");
                containerPadding = pxPosition === -1 ? null : Number(containerPadding.substr(0, pxPosition))

                logo.style.left = -(leftContainerOffset + logo.offsetWidth * 2 + (containerPadding ?? 20)) + "px";
            }
        }

        repositionSubmitLogo()
    }, [registrationContainer, logoSubmit])

    const handleSubmit = async (newUser: UserInformationInputInterface) => {
        console.log("[REGISTRATION] handleSubmit")
        if (!newUser.username) {
            handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: Please choose an username.")
        } else if (newUser.password !== newUser.passwordConfirmation) {
            handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: Password and password confirmation must match.")
        } else if (newUser.password.match(/\d/) == null || newUser.password.match(/[a-z]/) == null || newUser.password.match(/[A-Z]/) == null || notMinStrLength(newUser.password, MIN_PASSWORD_LENGTH)) {
            handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: Please pay attention to the notes below the input fields.")
        } else {
            registerNewUser(newUser.username, newUser.password, newUser.passwordConfirmation)
                .then(res => {
                    handleAlertVisibility(DEFAULT_ALERT_DURATION, "success", "Worked: " + (res.outputMessage ? res.outputMessage : (res.httpStatus + " " + res.httpMessage)));
                    toggleSubmitLogo();
                })
                .catch(err => {
                    handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: " + (err.outputMessage ? err.outputMessage : (err.httpStatus + " " + err.httpMessage)))
                })
        }
    }

    const handleAlertVisibility = (duration: number, color: "primary" | "secondary" | "success" | "danger" | "warning" | "info" | "light" | "dark", message: string) => {
        if (!alertVisibility) {
            setAlertMessage(message);
            setAlertColor(color);
            setAlertVisibility(true);
            setTimeout(() => {
                setAlertVisibility(false);
            }, duration);
        }
    }

    return (
        <Container id="registrationContainer">
            <Row>
                <div className="px-3 w-100">
                    <h1>Create new account</h1>
                    <UserInformationInput triggerAlert={handleAlertVisibility} submitFunction={handleSubmit}/>
                    <Alert variant={alertVariant} onClose={() => setAlertVisibility(false)} show={alertVisibility}
                           dismissible>
                        <p>{alertMessage}</p>
                    </Alert>
                </div>
            </Row>
            <img className={"invisible m0 position-relative"} src={fileFighter} alt="logo"
                 id="logoSubmit"/>
        </Container>
    )

    function toggleSubmitLogo() {
        const logo = document.getElementById("logoSubmit")
        if (logo) {
            const size: getWindowSize_Interface = getWindowSize();

            setTimeout(() => { //run right
                logo.style.transition = "4s";
                logo.classList.remove("invisible")
                logo.classList.add("visible")
                logo.style.transform = "translateX(" + (logo.offsetWidth + size.viewportWidth) + "px)";
            }, 1000);
            setTimeout(() => { //turn around
                logo.style.transition = "2s";
                logo.style.transform = "translateX(" + (logo.offsetWidth + size.viewportWidth) + "px) scaleX(-1)";
            }, 4000);
            setTimeout(() => { //run left
                logo.style.transition = "4s";
                logo.style.transform = "scaleX(-1)";
            }, 5000);
            setTimeout(() => { //turn around
                logo.style.transform = "";
                logo.classList.add("invisible")
                logo.classList.remove("visible")
            }, 8000);
        }
    }
}
