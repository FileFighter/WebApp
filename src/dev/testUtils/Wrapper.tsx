import React from "react";
import { BrowserRouter } from "react-router-dom";
import store from "../../background/redux/store";
import { Provider } from "react-redux";

export const Wrapper: React.FC = ({ children }) => {
    return (
        <Provider store={store}>
            <BrowserRouter>{children}</BrowserRouter>
        </Provider>
    );
};
