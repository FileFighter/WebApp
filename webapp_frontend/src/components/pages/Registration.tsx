import React, {ChangeEvent, FormEvent, ReactElement, useEffect, useState} from "react";
import {Container, Row, Col, Form, FormGroup, Button, Alert} from "react-bootstrap";
import {deleteSpaces} from "../../background/methods/strings";
import {biggerMaxStrLength, notMinStrLength} from "../../background/methods/checkInput";
import info_svg from "../../assets/images/icons/material.io/info-24px.svg";
import check_svg from "../../assets/images/icons/material.io/check_circle-24px.svg";
import error_svg from "../../assets/images/icons/material.io/error-24px.svg";
import fileFighter from "../../assets/images/logos/logo.png";
import {registerNewUser} from "../../background/api/registration";
import {getWindowSize, getWindowSize_Interface} from "../../background/methods/windowSize";
import {getStyleValue} from "../../background/methods/style";

export default function Registration(): ReactElement {
    const MIN_PASSWORD_LENGTH = 8;
    const MAX_PASSWORD_LENGTH = 20;
    const DEFAULT_ALERT_DURATION = 3500;

    const [username, setUsername] = useState<string>("");
    const [password, setPassword] = useState<string>("");
    const [passwordConfirmation, setPasswordConfirmation] = useState<string>("");
    const [passwordInformationLength, setPasswordInformationLength] = useState<boolean>(false);
    const [passwordInformationLowercase, setPasswordInformationLowercase] = useState<boolean>(false);
    const [passwordInformationUppercase, setPasswordInformationUppercase] = useState<boolean>(false);
    const [passwordInformationNumber, setPasswordInformationNumber] = useState<boolean>(false);
    const [passwordsMatch, setPasswordsMatch] = useState<boolean>(true);
    const [alertMessage, setAlertMessage] = useState<string>("Error 404: No Message found.");
    const [alertVariant, setAlertColor] = useState<"primary" | "secondary" | "success" | "danger" | "warning" | "info" | "light" | "dark">("success");
    const [alertVisibility, setAlertVisibility] = useState<boolean>(false);

    useEffect(() => {
        function repositionSubmitLogo() {
            const logo = document.getElementById("logoSubmit")
            if (logo){
                const container: HTMLElement | null = document.getElementById("registrationContainer");
                const leftContainerOffset: number = container?.getBoundingClientRect().left ?? 0;

                let containerPadding:string|number|null = getStyleValue(container, "padding-left");
                const pxPosition = containerPadding.indexOf("px");
                containerPadding = pxPosition == -1 ? null : Number(containerPadding.substr(0, pxPosition))

                logo.style.left = -(leftContainerOffset+logo.offsetWidth*2+(containerPadding ?? 20)) + "px";
            }
        }
        repositionSubmitLogo()
    },[document.getElementById("registrationContainer"), document.getElementById("logoSubmit")])

    useEffect(() => {
        reviewPasswordMatch()
        // eslint-disable-next-line
    }, [passwordConfirmation, password])

    const handleSubmit = async (event: FormEvent) => {
        console.log("[REGISTRATION] handleSubmit")
        event.preventDefault();
        reviewPasswordMatch();
        if (!username) {
            toggleSubmitLogo();
            //handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: Please choose an username.")
        } else if (!passwordsMatch) {
            handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: Password and password confirmation must match.")
        } else if (!passwordInformationNumber || !passwordInformationLowercase || !passwordInformationUppercase || !passwordInformationLength) {
            handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: Please pay attention to the notes below the input fields.")
        } else {
            await registerNewUser(username, password, passwordConfirmation)
                .then(res => {
                    handleAlertVisibility(DEFAULT_ALERT_DURATION, "success", "Worked: " + (res.outputMessage ? res.outputMessage : (res.httpStatus + " " + res.httpMessage)));
                    //toggleSubmitLogo();
                })
                .catch(err => {
                    handleAlertVisibility(DEFAULT_ALERT_DURATION, "danger", "Error: " + (err.outputMessage ? err.outputMessage : (err.httpStatus + " " + err.httpMessage)))
                })
        }
    }

    const handleAlertVisibility = (duration: number, color: "primary" | "secondary" | "success" | "danger" | "warning" | "info" | "light" | "dark", message: string) => {
        if (!alertVisibility) {
            setAlertMessage(message);
            setAlertColor(color);
            setAlertVisibility(true);
            setTimeout(() => {
                setAlertVisibility(false);
            }, duration);
        }
    }


    const makePasswordInputFitRules = (input: string): [string, boolean] => {
        input = deleteSpaces(input);
        if (biggerMaxStrLength(input, MAX_PASSWORD_LENGTH)) {
            handleAlertVisibility(DEFAULT_ALERT_DURATION, "warning", "Maximum password length exceeded. Input was undone.");
            return [input, false];
        }
        return [input, true];
    }

    const handlePasswordChange = (event: ChangeEvent<HTMLInputElement>) => {
        event.preventDefault();
        let value: [string, boolean] | string = makePasswordInputFitRules(event.target.value);
        if (!value[1]) {
            value = password;
        } else {
            value = value[0]
        }
        setPasswordInformationLength(!notMinStrLength(value, MIN_PASSWORD_LENGTH));
        setPasswordInformationLowercase(value.match(/[a-z]/) !== null);
        setPasswordInformationUppercase(value.match(/[A-Z]/) !== null);
        setPasswordInformationNumber(value.match(/\d/) !== null);
        setPassword(value)
    }

    const handlePasswordConfirmationChange = async (event: ChangeEvent<HTMLInputElement>) => {
        event.preventDefault();
        let value: [string, boolean] | string = makePasswordInputFitRules(event.target.value);
        if (!value[1]) {
            value = passwordConfirmation;
        } else {
            value = value[0]
        }
        setPasswordConfirmation(value);
    }

    const reviewPasswordMatch = (): void => {
        setPasswordsMatch(password === passwordConfirmation);
    }

    return (
        <Container className="h-100" style={{position: "relative"}} id="registrationContainer">
            <Row>
                <Col md={{span: 6, offset: 3}}>
                    <h1>Create new account</h1>
                    <Form onSubmit={handleSubmit}>
                        <FormGroup controlId="formBasicUsername">
                            <Form.Label>Username</Form.Label>
                            <Form.Control type={"name"} value={username}
                                          onChange={event => setUsername(event.target.value)}/>
                        </FormGroup>
                        <Form.Group controlId="formBasicPassword">
                            <Form.Label>Password</Form.Label>
                            <Form.Control type="password"
                                          placeholder="Must contain one number, uppercase & lowercase letter each"
                                          value={password}
                                          onChange={(event: ChangeEvent<HTMLInputElement>) => handlePasswordChange(event)}/>
                            <div>
                                <img alt={"status icon password length"}
                                     src={passwordInformationLength ? check_svg : info_svg}/>
                                <span className={"sr-only"}>{passwordInformationLength ? "Done: " : "Missing: "}</span>
                                <span className={passwordInformationLength ? "text-success" : "text-muted"}>Passwords must be between 8 and 20 characters.</span>
                            </div>
                            <div>
                                <img alt={"status icon password contains uppercase character"}
                                     src={passwordInformationUppercase ? check_svg : info_svg}/>
                                <span
                                    className={"sr-only"}>{passwordInformationUppercase ? "Done: " : "Missing: "}</span>
                                <span className={passwordInformationUppercase ? "text-success" : "text-muted"}>Passwords must be at least contain 1 uppercase character.</span>
                            </div>
                            <div>
                                <img alt={"status icon password contains lowercase character"}
                                     src={passwordInformationLowercase ? check_svg : info_svg}/>
                                <span
                                    className={"sr-only"}>{passwordInformationLowercase ? "Done: " : "Missing: "}</span>
                                <span className={passwordInformationLowercase ? "text-success" : "text-muted"}>Passwords must be at least contain 1 lowercase character.</span>
                            </div>
                            <div>
                                <img alt={"status icon password contains number"}
                                     src={passwordInformationNumber ? check_svg : info_svg}/>
                                <span className={"sr-only"}>{passwordInformationNumber ? "Done: " : "Missing: "}</span>
                                <span className={passwordInformationNumber ? "text-success" : "text-muted"}>Passwords must be at least contain 1 number.</span>
                            </div>
                        </Form.Group>
                        <Form.Group controlId="formConfirmPassword">
                            <Form.Label>Re-enter password</Form.Label>
                            <Form.Control type="password"
                                          value={passwordConfirmation}
                                          onChange={(event: ChangeEvent<HTMLInputElement>) => handlePasswordConfirmationChange(event)}/>
                            <div>
                                <img alt={"status icon passwords match"}
                                     src={!passwordConfirmation ? info_svg : passwordsMatch ? check_svg : error_svg}/>
                                <span className={"sr-only"}>{passwordsMatch ? "Done: " : "Missing: "}</span>
                                <span
                                    className={!passwordConfirmation ? "text-muted" : passwordsMatch ? "text-success" : "text-danger"}>Passwords must match.</span>
                            </div>
                        </Form.Group>
                        <Button variant="primary" type="submit">
                            Submit
                        </Button>
                        <Alert variant={alertVariant} onClose={() => setAlertVisibility(false)} show={alertVisibility}
                               dismissible>
                            <p>{alertMessage}</p>
                        </Alert>
                    </Form>
                </Col>
            </Row>
            <img style={{margin: 0, position: "relative", left: 0, visibility: "hidden"}} src={fileFighter} alt="logo"
                 id="logoSubmit"/>
        </Container>
    )

    function toggleSubmitLogo() {
        const logo = document.getElementById("logoSubmit")
        if (logo) {
            const size:getWindowSize_Interface = getWindowSize();

            setTimeout(() => { //run right
                logo.style.transition = "4s";
                logo.style.visibility = "visible";
                logo.style.transform = "translateX(" + (logo.offsetWidth + size.viewportWidth) + "px)";
            }, 1000);
            setTimeout(() => { //turn around
                logo.style.transition = "2s";
                logo.style.transform = "translateX(" + (logo.offsetWidth + size.viewportWidth) + "px) scaleX(-1)";
            }, 4000);
            setTimeout(() => { //run left
                logo.style.transition = "4s";
                logo.style.transform = "scaleX(-1)";
            }, 5000);
            setTimeout(() => { //turn around
                logo.style.transform = "";
                logo.style.visibility = "hidden";
            }, 8000);
        }
    }
}
