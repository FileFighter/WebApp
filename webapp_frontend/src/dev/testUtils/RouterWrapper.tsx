import React from "react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

export const RouterWrapper: React.FC = ({ children }) => {
  const history = createMemoryHistory();

  return <Router history={history}> {children} </Router>;
};
