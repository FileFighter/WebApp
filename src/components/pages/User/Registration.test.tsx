import { render } from "@testing-library/react"
import React from "react"
import Registration from "./Registration"
import { Wrapper } from "../../../dev/testUtils/Wrapper"

test("renders Login page", () => {
    const breadcrumb = render(
        <Wrapper>
            <Registration />
        </Wrapper>
    )

    expect(breadcrumb).toMatchSnapshot()
})
