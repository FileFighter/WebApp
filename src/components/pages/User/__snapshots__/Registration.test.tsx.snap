// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Login page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
       
      <div
        class="container"
        id="registrationContainer"
      >
        <div
          class="row"
        >
          <div
            class="px-3 w-100"
          >
            <h1>
              Create new account
            </h1>
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class="form-label"
                  for="formBasicUsername"
                >
                  Username
                </label>
                <input
                  class="form-control"
                  id="formBasicUsername"
                  type="name"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="form-label"
                  for="formBasicPassword"
                >
                  Password
                </label>
                <input
                  class="form-control"
                  id="formBasicPassword"
                  placeholder="Must contain one number, uppercase & lowercase letter each"
                  type="password"
                  value=""
                />
                <div>
                  <img
                    alt="status icon password length"
                    src="info-24px.svg"
                  />
                  <span
                    class="sr-only"
                  >
                    Missing: 
                  </span>
                  <span
                    class="text-muted"
                  >
                    Passwords must be between 8 and 20 characters.
                  </span>
                </div>
                <div>
                  <img
                    alt="status icon password contains uppercase character"
                    src="info-24px.svg"
                  />
                  <span
                    class="sr-only"
                  >
                    Missing: 
                  </span>
                  <span
                    class="text-muted"
                  >
                    Passwords must be at least contain 1 uppercase character.
                  </span>
                </div>
                <div>
                  <img
                    alt="status icon password contains lowercase character"
                    src="info-24px.svg"
                  />
                  <span
                    class="sr-only"
                  >
                    Missing: 
                  </span>
                  <span
                    class="text-muted"
                  >
                    Passwords must be at least contain 1 lowercase character.
                  </span>
                </div>
                <div>
                  <img
                    alt="status icon password contains number"
                    src="info-24px.svg"
                  />
                  <span
                    class="sr-only"
                  >
                    Missing: 
                  </span>
                  <span
                    class="text-muted"
                  >
                    Passwords must be at least contain 1 number.
                  </span>
                </div>
              </div>
              <div
                class="form-group"
              >
                <label
                  class="form-label"
                  for="formConfirmPassword"
                >
                  Re-enter password
                </label>
                <input
                  class="form-control"
                  id="formConfirmPassword"
                  type="password"
                  value=""
                />
                <div>
                  <img
                    alt="status icon passwords match"
                    src="info-24px.svg"
                  />
                  <span
                    class="sr-only"
                  >
                    Done: 
                  </span>
                  <span
                    class="text-muted"
                  >
                    Passwords must match.
                  </span>
                </div>
              </div>
              <button
                class="btn btn-primary"
                type="submit"
              >
                Submit
              </button>
            </form>
          </div>
        </div>
        <img
          alt="logo"
          class="invisible m0 position-relative"
          id="logoSubmit"
          src="adventurer-run.gif"
          style="left: -20px;"
        />
      </div>
       
    </div>
  </body>,
  "container": <div>
     
    <div
      class="container"
      id="registrationContainer"
    >
      <div
        class="row"
      >
        <div
          class="px-3 w-100"
        >
          <h1>
            Create new account
          </h1>
          <form
            class=""
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="formBasicUsername"
              >
                Username
              </label>
              <input
                class="form-control"
                id="formBasicUsername"
                type="name"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="formBasicPassword"
              >
                Password
              </label>
              <input
                class="form-control"
                id="formBasicPassword"
                placeholder="Must contain one number, uppercase & lowercase letter each"
                type="password"
                value=""
              />
              <div>
                <img
                  alt="status icon password length"
                  src="info-24px.svg"
                />
                <span
                  class="sr-only"
                >
                  Missing: 
                </span>
                <span
                  class="text-muted"
                >
                  Passwords must be between 8 and 20 characters.
                </span>
              </div>
              <div>
                <img
                  alt="status icon password contains uppercase character"
                  src="info-24px.svg"
                />
                <span
                  class="sr-only"
                >
                  Missing: 
                </span>
                <span
                  class="text-muted"
                >
                  Passwords must be at least contain 1 uppercase character.
                </span>
              </div>
              <div>
                <img
                  alt="status icon password contains lowercase character"
                  src="info-24px.svg"
                />
                <span
                  class="sr-only"
                >
                  Missing: 
                </span>
                <span
                  class="text-muted"
                >
                  Passwords must be at least contain 1 lowercase character.
                </span>
              </div>
              <div>
                <img
                  alt="status icon password contains number"
                  src="info-24px.svg"
                />
                <span
                  class="sr-only"
                >
                  Missing: 
                </span>
                <span
                  class="text-muted"
                >
                  Passwords must be at least contain 1 number.
                </span>
              </div>
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="formConfirmPassword"
              >
                Re-enter password
              </label>
              <input
                class="form-control"
                id="formConfirmPassword"
                type="password"
                value=""
              />
              <div>
                <img
                  alt="status icon passwords match"
                  src="info-24px.svg"
                />
                <span
                  class="sr-only"
                >
                  Done: 
                </span>
                <span
                  class="text-muted"
                >
                  Passwords must match.
                </span>
              </div>
            </div>
            <button
              class="btn btn-primary"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
      <img
        alt="logo"
        class="invisible m0 position-relative"
        id="logoSubmit"
        src="adventurer-run.gif"
        style="left: -20px;"
      />
    </div>
     
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
